compiler/next/include/chpl/AST/ASTClassesList.h

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// This file lists the AST classes to help with macros generating code for each
+//
+// Each line consists of one of
+//   AST_NODE(NAME)
+//   AST_LEAF(NAME)
+//   AST_BEGIN_SUBCLASSES(NAME)
+//   AST_END_SUBCLASSES(NAME)
+//
+// AST_NODE is for an AST node than can contain other AST nodes
+// AST_LEAF is for an AST node that cannot contain other AST nodes
+// AST_BEGIN_SUBCLASSES / AST_END_SUBCLASSES mark subclasses of abstract classes
+
+// These AST nodes are documented in their respective header files
+// and only listed here.
+
+// This file should store the AST nodes in groups according to the
+// class hierarchy and otherwise be in sorted order.
+
+// A note about the philosophy:
+// Where to draw the line on
+// new AST types vs. using a primitive?
+//  try to make AST nodes and use inheritance
+//  to allow coarse-grain view
+
+// Should there be a generic way to enumerate the
+// children of a particular AST node besides the visitor?
+//  yes, and allow getting "next" somehow
+//  let's try making a vector of children
+
+// the following comment disables doxygen for these
+/// \cond DO_NOT_DOCUMENT
+
+AST_BEGIN_SUBCLASSES(Expr)
+
+  AST_LEAF(Comment)
+  AST_LEAF(ErroneousExpr)
+  AST_LEAF(Identifier)
+
+  AST_NODE(IfExpr)
+  AST_NODE(TryExpr)
+  AST_NODE(DotExpr)
+
+  AST_BEGIN_SUBCLASSES(LoopExpr)
+    AST_NODE(ForExpr)
+    AST_NODE(ForallExpr)
+    AST_NODE(ForeachExpr)
+  AST_END_SUBCLASSES(LoopExpr)
+
+  AST_BEGIN_SUBCLASSES(LiteralExpr)
+    AST_LEAF(StringLiteral) // includes bytes
+    AST_LEAF(NumericLiteral)
+    AST_NODE(TupleLiteral)
+    AST_NODE(ArrayLiteral)
+    AST_NODE(DomainLiteral)
+  AST_END_SUBCLASSES(LiteralExpr)
+
+  AST_BEGIN_SUBCLASSES(CallExpr)
+    AST_NODE(PrimCallExpr)
+    AST_NODE(NewExpr)
+    AST_NODE(FnCallExpr)
+    AST_NODE(OpCallExpr)
```

Comment:
I am uneasy about `FnCallExpr` because it does not include paren-less calls. Maybe `ParenExpr` is a better name, similarly to DotExpr?

Following the lead of `NewExpr` and `ParenExpr`, I suggest simplifying out "Call" from the other names - changing to `PrimExpr` and `OpExpr`. "OpExpr" sounds to me better than "OpCallExpr" anyway.

Bummer that we can't put DotExpr under CallExpr. OTOH while that might be more elegant, the amount of factored out code if we did that would be next to zero.

Reply:
> Maybe `ParenExpr` is a better name?

Thinking some more, "ParenExpr" might imply a tuple literal. So not a good name for this class. How about `CallWithParens", or in any case something that has both "call" and "parens" ?

Reply:
I have made other changes here as well that I don't plan to merge in this PR (a lot of renaming). You can see what I have at the moment here: https://github.com/mppf/chapel/blob/new-asts-dev-3/compiler/next/include/chpl/AST/ASTClassesList.h

I have been removing the trailing `Expr` from these.

Are you saying that you'd like to have a different AST node for a paren-less call? I was imagining it would be a property on the call (e.g. a `bool hasParens` field/accessor).

Reply:
> Are you saying that you'd like to have a different AST node for a paren-less call? I was imagining it would be a property on the call (e.g. a bool hasParens field/accessor).

My observation is that we cannot have an AST node for paren-less calls. At parsing a paren-less call is either an Identifier or a Dot call. For example, given `A.b`, we do not know whether this is a call to `b` or a reference to a variable `b`. Likewise  in `if stridable then ...` "stridable" can be a variable or a call.

For this reason I consider the name `FnCall` to be too general, meaning it conceptually includes the calls that the class FnCall will not represent. So I propose `ParenCall` instead.

Reply:
> I am uneasy about FnCallExpr because it does not include paren-less calls.

So, the thing is, a paren-less call won't parse as a call at all. It is just an Identifier. E.g.

``` chapel
proc a { }
a;
```

the parser doesn't know the last statement is a call at all. We won't know that until resolution.

Hence the AST nodes should have nodes for things that are syntactically calls. Some of the Identifiers will end up meaning a call, but we can't give them a different node, because we don't know that yet.

So, I think Call/FnCall/etc does not need to be concerned about parenless calls. I don't think we need to rename it to ParenCall or something like that because we know these ones will be calls, even if some other things are also calls.